### PR TITLE
[Docs](mlu-ops):add bang-c-ops-develop-guide.

### PR DESCRIPTION
**Pull request by mahxn0**
Thanks for your contribution and we appreciate it a lot. :rocket::rocket:

## 1. Motivation

add bangc-develop-guide docs

## 2. Modification
new_file:   docs/BANGC-Develop-Guide.md
modified:   docs/Pull-Request.md
new file:   "docs/images/MLUOPS_\347\256\227\345\255\220\345\256\236\347\216\260.png"
renamed:    docs/PR-Createpr.png -> docs/images/PR-Createpr.png
renamed:    docs/PR-Fork.png -> docs/images/PR-Fork.png
new file:   docs/images/stride.png
new_file: docs/images/strideCompute.png
new file:   "docs/images/\344\272\224\347\272\247\346\265\201\346\260\264.png"

Are new test cases added? If so, please post the corresponding generator-PR link here.

## 3. Test Report

If you want to know how to do operator testing, you can see [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).

#### 3.1.1 Accuracy Acceptance Standard

For static threshold standard details, see: [MLU-OPS™ Accuracy Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Accuracy-Acceptance-Standard.md).

- static threshold
  - diff1
    - [ ] float32 mlu diff1 <= 1e-5
    - [ ] float32 mlu diff1 <= 3e-3
    - [ ] float16 mlu diff1 <= 3e-3
  - diff2
    - [ ] float32 mlu diff2 <= 1e-5
    - [ ] float32 mlu diff2 <= 3e-3
    - [ ] float16 mlu diff2 <= 3e-3
  - diff3
    - [ ] mlu diff3 == 0
    - [ ] mlu diff3_1 == 0
    - [ ] mlu diff3_2 == 0
- dynamic threshold
  - [ ] diff1: mlu diff1 <= max(baseline diff1 * 10, static threshold)
  - [ ] diff2: mlu diff2 <= max(baseline diff2 * 10, static threshold)
  - [ ] diff3: mlu diff3 <= max(baseline diff3 * 10, static threshold)
    - float32, threshold = 1e-5
    - float16, threshold = 1e-3

#### 3.1.2 Operator Scheme checklist

- Supported hardware
  - [ ] MLU370
  - [ ] MLU590
- Job types
  - [ ] BLOCK
  - [ ] UNION1
  - [ ] UNION2
  - [ ] UNION4
  - [ ] The operator will dynamically select the most suitable task type, for example, UNION8

#### 3.2.1 Accuracy Test

If you have checked the following items, please tick the relevant box.

- [ ] Data type test (e.g. float32/int8)
- [ ] Multi-dimensional tensor test
- [ ] Layout test
- [ ] Different size/integer remainder end segment/alignment misalignment test
- [ ] Zero dimensional tensor test/zero element test
- [ ] stability test
- [ ] Multiple platform test
- [ ] Gen_case module test, see: [Gencase-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/Gencase-User-Guide-zh.md)
- [ ] Nan/INF tests 
- [ ] Bug fix tests
- [ ] For memory leak check details, see: [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md)
- [ ] For code coverage check details, see: [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md)
- [ ] For I/O calculation efficiency check details, see: [MLU-OPS™-Performance-Acceptance-Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md)

#### 3.2.2 Parameter Check

Test Point-1: `When a new operator is submitted, the test points are given and the test results are stated`. Acceptance Standard: `Normal error`.
```bash
Please fill your test results(Error Message) in here, ...
```

Test Point-2: `Whether illegal parameters are passed`. Acceptance Standard: `Normal error`.
```bash
Test results...
```

### 3.3 Performance Test

See [MLU-OPS™ Performance Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md) for details.

Platform：MLU370

```bash
# The test results should contain Op name, Shape, Data type,  
#   MLU Hardware Time(us), MLU Interface Time(us), MLU IO Efficiency, 
#   MLU Compute Efficiency, and Mlu Workspace Size(Bytes)
# 
# for example:
#
# ----------- case0 -----------
# case0
# [Op name                ]: abs
# [Shape                  ]: input.shape=[1024,1024,3,4], output.shape=[1024,1024,3,4]
# [Data type]             ]: float32
# [MLU Hardware Time      ]: 15728 (us)
# [MLU Interface Time     ]: 369.008 (us)
# [MLU IO Efficiency      ]: 0.23275
# [MLU Compute Efficiency ]: 0.5
# [Mlu Workspace Size     ]: -1 (Bytes)
# 
# ----------- case1 -----------
# ...
```

Platform：MLU590
```bash
# ----------- case0 -----------
# ----------- case1 -----------
# ...
```

### 3.4 Summary Analysis

Please give a brief overview here, if you want to note and summarize the content.

### CODE REVIEW

File: docs/BANGC-Develop-Guide.md [hunk 1]
```diff
@@ -0,0 +1,3040 @@
...
+- 为用户提供一套完善的、可持续兼容的高层开发语言，实现性能优化
+- 支持单核或多核的、并行的编程模型
+
+具体指令使用介绍见[寒武纪官网开发者社区](https://developer.cambricon.com/index/document/index/classid/3.html)，算子开发章节[CAMBRICON BANG C 开发者手册](https://www.cambricon.com/docs/sdk_1.15.0/cntoolkit_3.7.2/cambricon_bang_c_4.7.2/index.html)
+
+
+
+## 2. 环境准备
+
+进行环境准备前，你需要了解如下基本概念，以便更好的理解后续操作。
```
Comment (PetrelYy) [inline docs/BANGC-Develop-Guide.md:18]: 查重率90%，hhhha，不知道这种内容上会不会侵权

File: docs/BANGC-Develop-Guide.md [hunk 2]
```diff
@@ -0,0 +1,3040 @@
...
+
+- 运行环境：指运行算子、推理程序、训练程序等的环境。运行环境必须带寒武纪MLU处理器。
+
+- 开发环境与运行环境合设场景：指带寒武纪MLU处理器的机器既作为开发环境又作为运行环境。此种场景下，代码开发与代码运行在同一台机器上。
+
+- 开发环境与运行环境分设场景：开发环境和运行环境不在同一台机器上，开发者使用带有寒武纪MLU处理器的机器作为运行环境；使用其他独立机器进行代码开发与编译，作为开发环境。
+
+  > 寒武纪MLU处理器: 若无特殊说明，以下均值寒武纪MLU370处理器。
+
+进行BANGC C算子开发之前，需要安装MLU驱动固件和CNToolkit软件包，完成开发环境和运行环境的准备
```
Comment (PetrelYy) [inline docs/BANGC-Develop-Guide.md:30]: 有些句子用了句号，有些没用，参考方都用了句号

File: docs/BANGC-Develop-Guide.md [hunk 3]
```diff
@@ -0,0 +1,3040 @@
...
+
+   ```
+   z = x + y
+   ```
+
+   计算逻辑是：将输入数据需要先搬运到AI Core的内部存储**NRAM**空间，使用计算接口完成两个输入参数相加，得到最终结果，再将输出搬出到外部存储GDRAM上。
+
+2. 明确输入和输出。
+
+   - Add算子有两个输入Tensor：x，y，一个输入变量：data_num，具体指相加的Tensor大小，输出为z。
```
Comment (PetrelYy) [inline docs/BANGC-Develop-Guide.md:76]: ：x、y

Comment (PetrelYy) [reply docs/BANGC-Develop-Guide.md:76]: 既然描述了输入类型，输出z也补充下
```suggestion
   - Add算子有两个输入Tensor：x、y，一个输入变量：data_num（Tensor a或 Tensor b 中元素个数），以及一个输出Tensor： z。
```

File: docs/BANGC-Develop-Guide.md [hunk 4]
```diff
@@ -0,0 +1,3040 @@
...
+
+   - 本样例中算子的输入支持的数据类型为float，算子输出的数据类型与输入数据类型相同。
+   - 算子输入支持shape（32，2048），输出shape与输入shape相同。
+   - 算子输入输出支持的format为：ARRAY。
```
Comment (PetrelYy) [inline docs/BANGC-Develop-Guide.md:80]: 参考中ND 表示两维度，描述上有些冗余， 咱们这里不照抄了
```suggestion
   - 算子输入输出 Tensor format 类型为 ARRAY，表示支持任意 shape。
```

File: docs/BANGC-Develop-Guide.md [hunk 5]
```diff
@@ -0,0 +1,3040 @@
...
+   ```
+
+   > 关于taskId，taskDim的介绍参考下述3.3.6 任务规模章节。
+
+4. 根据数据大小及计算公式，对nram进行拆分，并确定片上数据循环处理的整数段和余数段。
+
+   ```c++
+   int32_t align_num = NFU_ALIGN_SIZE / sizeof(float);
+     int32_t data_nram_num =
+       MAX_NRAM_SIZE / sizeof(float) / 2 / align_num * align_num;
```
Comment (PetrelYy) [inline docs/BANGC-Develop-Guide.md:236]: 错位了

File: docs/BANGC-Develop-Guide.md [hunk 6]
```diff
@@ -0,0 +1,3040 @@
...
+   int32_t rem_nram_num = data_per_core % data_nram_num;
+   ```
+
+   > nram空间拆分为2块，一份存放输入x，一份存放输入y，单次处理的数据按照128byte对齐。
+
+5. 整数段循环处理
+
+   ```c++
+   for (int32_t i = 0; i < loop_num; i++) {
+       __memcpy(a_nram, a_offset + i * data_nram_num,
```
Comment (PetrelYy) [inline docs/BANGC-Develop-Guide.md:249]: 5，6的代码段加一些注释吧

File: docs/BANGC-Develop-Guide.md [hunk 7]
```diff
@@ -0,0 +1,2946 @@
...
+  cd mlu-ops
+  git submodule update --init --recursive
+  ```
+
+- 准备CNToolkit、CNNL环境
+
+  具体的MLU-OPS™依赖版本可在[MLU-OPS™编译属性文件](https://github.com/Cambricon/mlu-ops/blob/master/build.property)中查看，将版本修改后执行下述命令完成环境下载
+
+  ```bash
+  wget https://sdk.cambricon.com/static/Basis/MLU370_X86_ubuntu20.04/cntoolkit_3.8.4-1.ubuntu20.04_amd64.deb
```
Comment (DanieeelLiu) [inline docs/BANGC-Develop-Guide.md:1988]: 这里的链接可能需要确认下，basis 这个路径可能已修改

File: docs/BANGC-Develop-Guide.md [hunk 8]
```diff
@@ -0,0 +1,2946 @@
...
+因此，对于一个在Memory上不连续的Tensor，算子处理的过程为：
+
+（1）在load过程中将其从**GDRAM上不连续变为NRAM上连续**（strideIn - gather）
+
+（2）在NRAM上完成连续tensor的片上计算
+
+（3）在store的过程中，将该NRAM上的结果从**NRAM上连续变为GDRAM上不连续**（strideOut - scatter）
+
+<p align="center">
+    <img src="./strideCompute.png" alt="example" style="zoom:50%">
```
Comment (DanieeelLiu) [inline docs/BANGC-Develop-Guide.md:2949]: 要考虑侵权风险，可能不是内部自己画的

File: docs/BANGC-Develop-Guide.md [hunk 9]
```diff
@@ -0,0 +1,2946 @@
...
+- ptr：指定该tensor在内存上的起始位置
+
+- shape：指定该tensor每个维度在内存上取多少个数
+
+- stride：指定该tensor每个维度在内存上取数时，每次跳多少个数
+
+那么，给定一段memory，一个tensor的ptr为该memory初始位置，shape=[2, 3, 2]，stride = [16, 5, 2]. 最终这个tensor为？
+
+<p align="center">
+    <img src="./stride.png" alt="example" style="zoom:50%">
```
Comment (DanieeelLiu) [inline docs/BANGC-Develop-Guide.md:2909]: 要考虑有侵权风险，可能不是内部自己画的

File: docs/BANGC-Develop-Guide.md [hunk 10]
```diff
@@ -0,0 +1,2952 @@
...
+## 1. BANG C简介
+
+BANG C语言是寒武纪针对MLU ( Machine Learning Unit )硬件提出的编程语言。
+
+- 提供高效的编程接口，充分发挥寒武纪硬件特性
+- 提供一种通用的异构编程方式，方便用户拓展自己的应用程序
+- 为用户提供一套完善的、可持续兼容的高层开发语言，实现性能优化
+- 支持单核或多核的、并行的编程模型
+
+具体指令使用介绍见[寒武纪官网开发者社区](https://developer.cambricon.com/index/document/index/classid/3.html)，算子开发章节[CAMBRICON BANG C 开发者手册](https://www.cambricon.com/docs/sdk_1.15.0/cntoolkit_3.7.2/cambricon_bang_c_4.7.2/index.html)
```
Comment (PetrelYy) [inline docs/BANGC-Develop-Guide.md:12]: 统一句号或都不写，整篇文档过一下

File: docs/BANGC-Develop-Guide.md [hunk 11]
```diff
@@ -0,0 +1,2952 @@
...
+
+
+
+## 2. 环境准备
+
+说明：运行环境需带寒武纪MLU处理器。
+
+### 驱动安装
+
+寒武纪发布的驱动安装包有两种：deb包和rpm包。寒武纪驱动安装包使用dkms框架来管理驱动的安装、升级和卸载。dkms可以针对系统内安装的每个内核版本，通过编译、安装相应的内核模块来实现驱动的加载。用户只需安装、更新相应的dkms软件包（deb包或rpm包）即可实现驱动的安装和升级。目前驱动已经支持主流Linux发行版系统（CentOS/Ubuntu/Debian/Kylin/UOS/Rocky Linux/OpenEuler/Anolis/CTyunOS/BCLinux/Linx）。
```
Comment (PetrelYy) [inline docs/BANGC-Develop-Guide.md:22]: 中英文之间加个空格，如
```suggestion
寒武纪发布的驱动安装包有两种：deb 包和 rpm 包。寒武纪驱动安装包使用 dkms 框架来管
```

File: docs/BANGC-Develop-Guide.md [hunk 12]
```diff
@@ -0,0 +1,2952 @@
...
+CNToolkit软件包安装步骤与介绍可参见 《[寒武纪CNToolkit安装升级使用手册](https://www.cambricon.com/docs/sdk_1.15.0/cntoolkit_3.7.2/cntoolkit_install_3.7.2/index.html)》
+
+Driver/CNToolKit 为必装依赖项，对于其它组件，用户可根据框架或安装包的依赖需求，自行选择匹配的组件库进行安装。可在[SDK下载](https://sdk.cambricon.com/download?sdk_version=V1.15.0&component_name=Neuware+SDK)界面下载对应的版本安装，本章节不再单独介绍。
+
+
+## 3. 算子开发（基础篇）
+
+### 3.1 使用说明
+
+算子(Operator,简称 OP)是深度学习算法的基本函数计算单元。 在网络模型中，算子对应网络层中的计算逻辑。
```
Comment (PetrelYy) [inline docs/BANGC-Develop-Guide.md:43]:
```suggestion
算子(Operator，简称 OP)是深度学习算法的基本函数计算单元。 在网络模型中，算子对应网络层中的计算逻辑。
```

File: docs/BANGC-Develop-Guide.md [hunk 13]
```diff
@@ -0,0 +1,2952 @@
...
+
+对于初次接触BANG-C的开发者，从本章节开始，我们以Add算子为例，介绍算子开发的基本步骤，主要包括算子分析、Kernel代码开发、Host代码开发、编译、测试和优化。同时引入BANG-C抽象硬件模型，BANGC-C抽象存储模型，BANG-C编程模型，BANG-C并发与同步，带你深入了解基于MLU硬件的BANG-C算子开发。
+
+### 3.2 快速入门
+
+在正式的开发之前，需要先按照第二章节所述完成**环境准备**工作。
```
Comment (PetrelYy) [inline docs/BANGC-Develop-Guide.md:49]: 这里的环境准备，可以放章节链接，比如
https://github.com/Cambricon/mlu-ops/blob/master/docs/Gencase-User-Guide-zh.md#2-%E7%AE%97%E5%AD%90%E4%B8%AD%E6%B7%BB%E5%8A%A0-gen_case-%E5%8A%9F%E8%83%BD

File: docs/BANGC-Develop-Guide.md [hunk 14]
```diff
@@ -0,0 +1,2952 @@
...
+
+#### 3.2.1 算子分析
+
+首先需要确定算子的功能、输入/输出和逻辑表达，其中重点分析算子算法的原理，提取出算子的数学表达式，再查询需要调用的[BANG C指令](https://www.cambricon.com/docs/sdk_1.15.0/cntoolkit_3.7.2/cambricon_bang_c_4.7.2/index.html)，确认指令介绍能否满足算子实现的要求。若无法满足算子实现的要求，可考虑对数学表达式变形或用多条指令拼凑出需要的指令功能来实现算子功能。
```
Comment (PetrelYy) [inline docs/BANGC-Develop-Guide.md:53]: 算子算法的原理 /// 是不是改成 “算子原理” 或者 “算子计算逻辑”更好？

File: docs/BANGC-Develop-Guide.md [hunk 15]
```diff
@@ -0,0 +1,2952 @@
...
+   ```
+
+   计算逻辑：即对输入的两个Tensor逐元素相加，输出相加后的Tensor。
+
+2. 确认输入输出相关参数
+
+   - 输入输出Tensor：Tensor x、Tensor y、Tensor z。
+   - 输入输出Layout：本样例中以ARRAY为例进行开发，表示支持任意数据类型。
+   - 输入输出Shape：本样例中以shape（32，2048）为例，输出shape与输入shape相同。
+   - 输入输出DType: 本样例中以float为例，输出与输入数据类型一致。
```
Comment (PetrelYy) [inline docs/BANGC-Develop-Guide.md:70]: 这里的冒号是英文冒号

File: docs/BANGC-Develop-Guide.md [hunk 16]
```diff
@@ -0,0 +1,2952 @@
...
+
+3. 确认算子实现所需接口。
+
+   - 数据搬运接口
+
+     从cpu ddr内存搬运到gdram内存，查阅[CNRT手册](https://www.cambricon.com/docs/sdk_1.15.0/cntoolkit_3.7.2/cnrt_6.7.0/index.html)，确认调用接口：[cnrtMemcpy](https://www.cambricon.com/docs/sdk_1.15.0/cntoolkit_3.7.2/cnrt_6.7.0/h2rst_generated/memory_management.html#cnrtmemcpy)。
+
+     从gdram内存搬运到nram内存，查阅[BANG C手册](https://www.cambricon.com/docs/sdk_1.15.0/cntoolkit_3.7.2/cambricon_bang_c_4.7.2/index.html)，确认调用接口：[__memcpy](https://www.cambricon.com/docs/sdk_1.15.0/cntoolkit_3.7.2/cambricon_bang_c_4.7.2/2Builtin-Functions/1D%20Memcpy%20Functions.html#memcpy)。
+
+     关于nram，gdram内存介绍可参考3.3.2章节。
```
Comment (PetrelYy) [inline docs/BANGC-Develop-Guide.md:80]:
```suggestion
     nram、gdram 介绍可参考 3.3.2 章节。
```

File: docs/BANGC-Develop-Guide.md [hunk 17]
```diff
@@ -0,0 +1,2952 @@
...
+
+   - 计算接口：初步分析可使用[__bang_add](https://www.cambricon.com/docs/sdk_1.15.0/cntoolkit_3.7.2/cambricon_bang_c_4.7.2/2Builtin-Functions/Vector%20Operation%20Functions.html#bang-add)实现。
+
+   - 内存管理接口：使用[cnrtMalloc](https://www.cambricon.com/docs/sdk_1.15.0/cntoolkit_3.7.2/cnrt_6.7.0/h2rst_generated/memory_management.html#cnrtmalloc)，[cnrtFree](https://www.cambricon.com/docs/sdk_1.15.0/cntoolkit_3.7.2/cnrt_6.7.0/h2rst_generated/memory_management.html#cnrtfree)进行申请和释放。
```
Comment (PetrelYy) [inline docs/BANGC-Develop-Guide.md:84]: 逗号顿号可以整篇check遍

File: docs/BANGC-Develop-Guide.md [hunk 18]
```diff
@@ -0,0 +1,2952 @@
...
+> NEUWARE_HOME为cntoolkit安装目录
+
+以下样例代码在samples/BANG/3_Performance/vectoradd/4_vectorAdd_multicore/vectorAdd_multicore.mlu中实现。
+
+1. 首先根据[核函数定义](https://www.cambricon.com/docs/sdk_1.15.0/cntoolkit_3.7.2/programming_guide_1.7.0/programming_model/index.html#kernel)中介绍的规则进行核函数的定义。
+
+   ```c++
+   __mlu_global__ void kernel(float *output, const float *a,
+                              const float *b, int32_t data_num)
+   ```
```
Comment (PetrelYy) [inline docs/BANGC-Develop-Guide.md:123]: 1. google c++风格，output 放最后，全局使用的变量记得是放前
2. 前文中使用的变量是 xyz，这里成了 ab outout
3. 接口每个参数一行吧，看起来更整洁

```suggestion
   ```c++
   __mlu_global__ void kernel(int32_t data_num,
                              const float *x,
                              const float *y,
                              float *output)
   ```
```

File: docs/BANGC-Develop-Guide.md [hunk 19]
```diff
@@ -0,0 +1,2952 @@
...
+
+4. 根据数据大小及计算公式，对nram进行拆分，并确定片上数据循环处理的整数段和余数段。
+
+   ```c++
+   int32_t align_num = NFU_ALIGN_SIZE / sizeof(float);
+   int32_t data_nram_num =  MAX_NRAM_SIZE / sizeof(float) / 2 / align_num *       align_num;
+   float *a_nram = (float *)nram_buffer;
+   float *b_nram = (float *)a_nram + data_nram_num;
+   int32_t loop_num = data_per_core / data_nram_num;
+   int32_t rem_nram_num = data_per_core % data_nram_num;
```
Comment (PetrelYy) [inline docs/BANGC-Develop-Guide.md:158]:
```suggestion
   const int32_t align_num = NFU_ALIGN_SIZE / sizeof(float);
   const int32_t data_nram_num =  MAX_NRAM_SIZE / sizeof(float) / 2 / align_num * align_num;
   const int32_t loop_num = data_per_core / data_nram_num;
   const int32_t rem_nram_num = data_per_core % data_nram_num;
   
   float *a_nram = (float *)nram_buffer;
   float *b_nram = (float *)a_nram + data_nram_num;
```

File: docs/BANGC-Develop-Guide.md [hunk 20]
```diff
@@ -0,0 +1,2952 @@
...
+   ```
+   > nram空间拆分为2块，一份存放输入x，一份存放输入y，单次处理的数据按照128byte对齐。
+
+5. 整数段循环处理
+
+   ```c++
+    for (int32_t i = 0; i < loop_num; i++) {
+      // Load data from GDRAM2NRAM
+      __memcpy(a_nram, a_offset + i * data_nram_num,
```
Comment (PetrelYy) [inline docs/BANGC-Develop-Guide.md:167]: 这里缩进显示有问题
顺便不换行了吧

File: docs/BANGC-Develop-Guide.md [hunk 21]
```diff
@@ -0,0 +1,3040 @@
+# BANG C算子开发
+
+## 1. BANG C简介
+
+BANG C语言是是寒武纪针对MLU ( Machine Learning Unit )硬件提出的编程语言
```
[the remark this answers is not in the document]
Comment (mahxn0) [reply docs/BANGC-Develop-Guide.md]: 已修改

File: docs/BANGC-Develop-Guide.md (hunk 1 above)
Comment (mahxn0) [reply docs/BANGC-Develop-Guide.md]: 已修改

File: docs/BANGC-Develop-Guide.md (hunk 2 above)
Comment (mahxn0) [reply docs/BANGC-Develop-Guide.md]: 已修改

File: docs/BANGC-Develop-Guide.md (hunk 3 above)
Comment (mahxn0) [reply docs/BANGC-Develop-Guide.md]: 已修改

File: docs/BANGC-Develop-Guide.md (hunk 4 above)
Comment (mahxn0) [reply docs/BANGC-Develop-Guide.md]: 已修改

File: docs/BANGC-Develop-Guide.md [hunk 22]
```diff
@@ -0,0 +1,2952 @@
...
+    __bang_add(a_nram, a_nram, b_nram, rem_align_num);
+    // Store data from NRAM2GDRAM
+    __memcpy(output_offset + loop_num * data_nram_num, a_nram,
+             rem_nram_num * sizeof(float), NRAM2GDRAM);
+   }
+   ```
+
+##### Host代码开发
+
+1. 创建任务[队列](https://www.cambricon.com/docs/sdk_1.15.0/cntoolkit_3.7.2/programming_guide_1.7.0/programming_interface/runtime/runtime.html#queue)
```
Comment (PetrelYy) [inline docs/BANGC-Develop-Guide.md:252]: 这里链接挺多，能否在哪里记录下文档版本，比如 toolkit=3.7.2,  文档处链接是拼接出来的。这样后续文档更新不用一条条更新？

File: docs/BANGC-Develop-Guide.md [hunk 23]
```diff
@@ -0,0 +1,2952 @@
...
+   CNRT_CHECK(cnrtPlaceNotifier(notifier_end, queue));
+   ```
+
+7. 同步等待Kernel计算完成后，将Device内存拷贝到Host
+
+   ```c++
+   CNRT_CHECK(cnrtPlaceNotifier(notifier_end, queue));
+   CNRT_CHECK(cnrtQueueSync(queue));
+   CNRT_CHECK(cnrtMemcpy(dst_cpu, dst_mlu, ELEM_NUM * sizeof(float),
+                       cnrtMemcpyDevToHost));
```
Comment (PetrelYy) [inline docs/BANGC-Develop-Guide.md:309]: 这里对齐错了

File: docs/BANGC-Develop-Guide.md [hunk 24]
```diff
@@ -0,0 +1,2952 @@
...
+             rem_nram_num * sizeof(float), GDRAM2NRAM);
+    __memcpy(b_nram, b + loop_num * data_nram_num,
+             rem_nram_num * sizeof(float), GDRAM2NRAM);
+    __bang_add(a_nram, a_nram, b_nram, rem_align_num);
+    __memcpy(output + loop_num * data_nram_num, a_nram,
+           rem_nram_num * sizeof(float), NRAM2GDRAM);
+  }
+}
+
+int main() {
```
Comment (PetrelYy) [inline docs/BANGC-Develop-Guide.md:559]: 这里host,device 代码放一起了，但是章节说明里却分了 host,device

File: docs/BANGC-Develop-Guide.md [hunk 25]
```diff
@@ -0,0 +1,2952 @@
...
+
+以上例子是简单的Block任务，任务规模为（1,1,1），仅使用一个core来完成既定大小Tensor的绝对值运算的单核向量化版本，对于如何使用多个core、多任务并行、大规模数据拆分及排流水指令并行，我们将在算子开发进阶篇为您逐步揭开。进入进阶篇之前，学习以下硬件架构，存储模型，指令计算并行等内容是必要的。
+
+
+
+#### 3.3 基本概念
+
+##### 3.3.1 抽象硬件模型
+
+<p align="center">
```
Comment (PetrelYy) [inline docs/BANGC-Develop-Guide.md:625]: 图挂了

File: docs/BANGC-Develop-Guide.md [hunk 26]
```diff
@@ -0,0 +1,3001 @@
...
+    if (i < n) {
+      L(i);
+    }
+    __sync_cluster();
+  }
+  ```
+
+  我们采用模板 2 实现 Add 算子四级流水，`vectoradd_common.h`同多核优化，Kernel/Host 完整代码如下：
+
+  ```
```
Comment (PetrelYy) [inline docs/BANGC-Develop-Guide.md:1383]: 没加c++

File: docs/BANGC-Develop-Guide.md [hunk 27]
```diff
@@ -0,0 +1,3001 @@
...
+  S(n-2);
+  C(n-1);
+  __sync_io_move_compute();
+  S(n-1);
+  __sync_io();
+  ```
+
+  实现模板二：
+
+  ```
```
Comment (PetrelYy) [inline docs/BANGC-Develop-Guide.md:1362]: 这里

File: docs/BANGC-Develop-Guide.md (hunk 24 above)
Comment (PetrelYy) [reply docs/BANGC-Develop-Guide.md:551]: 既然放一个文件里，那就在 main 前加个注释说明下面的是host侧

**Review by PetrelYy — APPROVED** [review]

**Review by DanieeelLiu — APPROVED** [review]

**Review by duzekunKTH — APPROVED** [review]

**Pull request merged**